app/test/circuit_breaker_test.go

```diff
@@ -0,0 +1,102 @@
+	coretypes "github.com/tendermint/tendermint/types"
+)
+
+const (
+	granter      = "granter"
+	grantee      = "grantee"
+	appVersion   = v1.Version
+	amountToSend = 1
+)
+
+var (
+	expiration = time.Now().Add(time.Hour)
+)
+
+// TestCircuitBreaker verifies that the circuit breaker prevents a nested Authz
+// message that contains a MsgTryUpgrade if the MsgTryUpgrade is not supported
+// in the current version.
+func TestCircuitBreaker(t *testing.T) {
+	config := encoding.MakeConfig(app.ModuleEncodingRegisters...)
+	testApp, keyRing := util.SetupTestAppWithGenesisValSet(app.DefaultInitialConsensusParams(), granter, grantee)
+	header := tmproto.Header{Height: 2, Version: version.Consensus{App: appVersion}}
+
+	signer, err := user.NewSigner(keyRing, config.TxConfig, testutil.ChainID, appVersion, user.NewAccount(granter, 1, 0))
+	require.NoError(t, err)
+
+	granterAddress := getAddress(t, granter, keyRing)
+	granteeAddress := getAddress(t, grantee, keyRing)
+
+	authorization := authz.NewGenericAuthorization(signaltypes.URLMsgTryUpgrade)
+	msg, err := authz.NewMsgGrant(granterAddress, granteeAddress, authorization, &expiration)
+	require.NoError(t, err)
+	ctx := testApp.NewContext(true, header)
+	_, err = testApp.AuthzKeeper.Grant(ctx, msg)
+	assert.Error(t, err)
+	assert.ErrorContains(t, err, "/celestia.signal.v1.Msg/TryUpgrade doesn't exist.: invalid type")
+
+	testApp.BeginBlocker(ctx, abci.RequestBeginBlock{Header: header})
+
+	tryUpgradeTx := newTryUpgradeTx(t, signer, granterAddress)
+	res := testApp.DeliverTx(abci.RequestDeliverTx{Tx: tryUpgradeTx})
+	assert.Equal(t, uint32(0x25), res.Code, res.Log)
+	assert.Contains(t, res.Log, "message type /celestia.signal.v1.MsgTryUpgrade is not supported in version 1: feature not supported")
+
+	nestedTx := newNestedTx(t, signer, granterAddress)
+	res = testApp.DeliverTx(abci.RequestDeliverTx{Tx: nestedTx})
+	assert.Equal(t, uint32(0x1), res.Code, res.Log)
+	assert.Contains(t, res.Log, "circuit breaker disables execution of this message: /celestia.signal.v1.MsgTryUpgrade")
+}
+
+func newTryUpgradeTx(t *testing.T, signer *user.Signer, senderAddress sdk.AccAddress) coretypes.Tx {
+	msg := signaltypes.NewMsgTryUpgrade(senderAddress)
+	options := blobfactory.FeeTxOpts(1e9)
+
+	rawTx, err := signer.CreateTx([]sdk.Msg{msg}, options...)
+	require.NoError(t, err)
+
+	return rawTx
+}
+
+func newNestedTx(t *testing.T, signer *user.Signer, granterAddress sdk.AccAddress) coretypes.Tx {
+	innerMsg := signaltypes.NewMsgTryUpgrade(granterAddress)
+	msg := authz.NewMsgExec(granterAddress, []sdk.Msg{innerMsg})
+
+	options := blobfactory.FeeTxOpts(1e9)
+
+	rawTx, err := signer.CreateTx([]sdk.Msg{&msg}, options...)
+	require.NoError(t, err)
+
+	return rawTx
+}
+
+func getAddress(t *testing.T, account string, keyRing keyring.Keyring) sdk.AccAddress {
+	record, err := keyRing.Key(account)
+	require.NoError(t, err)
+
+	address, err := record.GetAddress()
+	require.NoError(t, err)
+
+	return address
+}
```

Comment:
nit: I think we might already have a function for this in testnode

Reply:
yeah we do https://github.com/celestiaorg/celestia-app/blob/2e49d665b3275e769dfe0371b1ca41e39dc3f5f5/test/util/testfactory/common.go#L83

Reply:
https://github.com/celestiaorg/celestia-app/pull/3554/commits/69b10599cb18caa353203a3047260abfedb29bc2